src/core/renderer_gl/renderer_gl.cpp

```diff
@@ -282,6 +437,50 @@ void Renderer::setupBlending() {
 	}
 }
 
+void Renderer::setupTextureEnvState() {
+	// TODO: Only update uniforms when the TEV config changed. Use an UBO potentially.
+
+	static constexpr std::array<u32, 6> ioBases = {
+	  0xc0, 0xc8, 0xd0, 0xd8, 0xf0, 0xf8
+	};
+
+	u32 textureEnvSourceRegs[6];
+	u32 textureEnvOperandRegs[6];
+	u32 textureEnvCombinerRegs[6];
+	float textureEnvColourRegs[6][4];
+	u32 textureEnvScaleRegs[6];
+
+	for (int i = 0; i < 6; i++) {
+		const u32 ioBase = ioBases[i];
+
+		textureEnvSourceRegs[i] = regs[ioBase];
+		textureEnvOperandRegs[i] = regs[ioBase + 1];
+		textureEnvCombinerRegs[i] = regs[ioBase + 2];
+
+		const u32 constantColor = regs[ioBase + 3];
+		textureEnvColourRegs[i][0] = (float)(constantColor & 0xff) / 255.0f;
```

Comment:
Maybe better off done in the FS?

Reply:
I ended up doing this in the VS and passing the result to the FS